libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,567 @@
+#include <range/v3/view/take_last.hpp>
+#include <range/v3/view/transform.hpp>
+
+using namespace solidity;
+using namespace solidity::yul;
+using namespace std;
+
+StackLayoutGenerator::StackLayoutGenerator(StackLayout& _layout): m_layout(_layout)
+{
+}
+
+namespace
+{
+struct PreviousSlot { size_t slot; };
+
+// TODO: Rewrite this as custom algorithm matching createStackLayout exactly and make it work
+// for all cases, including duplicates and removals of slots that can be generated on the fly, etc.
+// After that the util::permute* functions can be removed.
+Stack createIdealLayout(Stack const& _post, vector<variant<PreviousSlot, set<unsigned>>> layout)
+{
+	util::permuteDup(static_cast<unsigned>(layout.size()), [&](unsigned _i) -> set<unsigned> {
+		// For call return values the target position is known.
+		if (set<unsigned>* pos = get_if<set<unsigned>>(&layout.at(_i)))
+			return *pos;
+		// Previous arguments can stay where they are.
+		return {_i};
+	}, [&](unsigned _i) {
+		std::swap(layout.back(), layout.at(layout.size() - _i - 1));
+	}, [&](unsigned _i) {
+		auto positions = get_if<set<unsigned>>(&layout.at(layout.size() - _i));
+		yulAssert(positions, "");
+		if (positions->count(static_cast<unsigned>(layout.size())))
+		{
+			positions->erase(static_cast<unsigned>(layout.size()));
+			layout.emplace_back(set<unsigned>{static_cast<unsigned>(layout.size())});
+		}
+		else
+		{
+			optional<unsigned> duppingOffset;
+			for (unsigned pos: *positions)
+			{
+				if (pos != layout.size() - _i)
+				{
+					duppingOffset = pos;
+					break;
+				}
+			}
+			yulAssert(duppingOffset, "");
+			positions->erase(*duppingOffset);
+			layout.emplace_back(set<unsigned>{*duppingOffset});
+		}
+	}, [&]() {
+		yulAssert(false, "");
+	}, [&]() {
+		layout.pop_back();
+	});
+
+	// Now we can construct the ideal layout before the operation.
+	// "layout" has the declared variables in the desired position and
+	// for any PreviousSlot{x}, x yields the ideal place of the slot before the declaration.
+	vector<optional<StackSlot>> idealLayout(_post.size(), nullopt);
+	for (auto const& [slot, idealPosition]: ranges::zip_view(_post, layout))
+		if (PreviousSlot* previousSlot = std::get_if<PreviousSlot>(&idealPosition))
+			idealLayout.at(previousSlot->slot) = slot;
+
+	while (!idealLayout.empty() && !idealLayout.back())
+		idealLayout.pop_back();
+
+	return idealLayout | ranges::views::transform([](optional<StackSlot> s) {
+		yulAssert(s, "");
+		return *s;
+	}) | ranges::to<Stack>;
+}
+}
+
+Stack StackLayoutGenerator::propagateStackThroughOperation(Stack _exitStack, CFG::Operation const& _operation)
+{
+	Stack& stack = _exitStack;
+
+	vector<set<unsigned>> targetPositions(_operation.output.size(), set<unsigned>{});
+	size_t numToKeep = 0;
+	for (size_t idx: ranges::views::iota(0u, targetPositions.size()))
+		for (unsigned offset: findAllOffsets(stack, _operation.output.at(idx)))
+		{
+			targetPositions[idx].emplace(offset);
+			++numToKeep;
+		}
```

Comment:
```suggestion
		{
			targetPositions[idx] = findAllOffsets(stack, _operation.output.at(idx);
			numToKeep += targetPositions[idx].size();
		}
```
Something like that?